### ISSUE TITLE
clarify use of generation time and units of rates

### ISSUE THREAD

**Issue by petrelharp** [popsim-consortium/stdpopsim#1197]
In checking a QC, I just got mixed up about whether `mutation_rate` and `recombination_rate` are in units of "per year" or "per generation". As far as I can tell:

- mutation and recombination rates are in units of "per generation"
- time units in the "parameters" of demographic models say whether they are in years or generations, helpfully
- the msprime engine simulates things with time units of generations
- and so demographic models are *stored* with time units of generations ([e.g., here](https://github.com/popsim-consortium/stdpopsim/blob/main/stdpopsim/catalog/HomSap/demographic_models.py#L125))

However, the [SLiM engine does a bunch of things with the generation time](https://github.com/popsim-consortium/stdpopsim/blob/main/stdpopsim/slim_engine.py#L828), and so it's not actually clear to me that the SLiM engine is doing the right thing here, given that time units coming in are in units of generations?

At least, this needs to be documented better (i.e., to say the units of these rates in the places we mention them, and clarify in the "adding a species" docs); I'm happy to do this. But, do you remember what's going on with the SLiM engine, @grahamgower?

**Comment by grahamgower** [comment]
Everything is per generation. The SLiM engine attempts to output a faithful representation of the demographic model into the slim script (as multidimensional arrays), with times given in the same units as the original model. When the slim script runs, it uses the provided generation time to convert times back into generations.

My intention was for a user to be able to grab the generated slim script and use this in isolation (potentially with modifications).  I still think this was a useful design goal, but probably the resulting complexity of the slim script is too much of a barrier for this in practice.

**Comment by petrelharp** [comment]
> Everything is per generation. The SLiM engine attempts to output a faithful representation of the demographic model into the slim script (as multidimensional arrays), with times given in the same units as the original model. When the slim script runs, it uses the provided generation time to convert times back into generations.

Hold on, I'm confused (apologies). Here you're implying that some demographic models *don't* have times in generations. Can you point to an example?

**Comment by grahamgower** [comment]
Sorry, yes, the demographic models are stored with time units in generations (as one expects for an msprime.Demography object). But most code implementing a demographic model (i.e. the human-readable implementation) writes time values in years for interpretability and/or easy comparison to a publication, and subsequently converts them to generations as the model is constructed.

**Comment by petrelharp** [comment]
Sorry, I'm still being dense - it sounds like everything is OK, but: how do you tell if a model is written in years or generations? And where does "writes time values in years" go?

**Comment by grahamgower** [comment]
What I mean is that when we look at the code implementing demographic models, we mostly see things like this:
https://github.com/popsim-consortium/stdpopsim/blob/33d5e2dce84ace9e1168f82c8ffc99e08c0f1bde/stdpopsim/catalog/HomSap/demographic_models.py#L275-L279

In the SLiM engine, there really is no way to know if the model was written like that. Some demographic models do use generations as units (e.g. the Zigzag model). So you're right that this is all messy and poorly defined. When we convert all implementations to Demes, all this will be more explicit and clear I think.

**Comment by petrelharp** [comment]
Ok, I got that part. But then what's going on with all the conversions with `generation_time` in the slim engine code? Since times are already in generations, why do they need to be converted?

**Comment by grahamgower** [comment]
They don't need to be converted. It is redundant.

**Comment by grahamgower** [comment]
Let me try to explain again...

Because model implementations looked like this:
https://github.com/popsim-consortium/stdpopsim/blob/33d5e2dce84ace9e1168f82c8ffc99e08c0f1bde/stdpopsim/catalog/HomSap/demographic_models.py#L275-L279

I wanted the generated SLiM scripts to have values in years like `220e3` inside the multidimensional arrays. At the time I wrote the code there were, maybe, 5 demographic models in stdpopsim.

So the SLiM engine has a redundant conversion to years, then back to generations.

**Comment by petrelharp** [comment]
I'd still like to go clarify the documentation to say what units the rates are in.
